File: src/test/java/org/cloudfoundry/credhub/controller/v1/KeyUsageControllerTest.java
package org.cloudfoundry.credhub.controller.v1;

import org.cloudfoundry.credhub.data.CredentialVersionDataService;
import org.cloudfoundry.credhub.data.EncryptionKeyCanaryDataService;
import org.cloudfoundry.credhub.service.EncryptionKey;
import org.cloudfoundry.credhub.service.EncryptionKeySet;
import org.cloudfoundry.credhub.service.EncryptionService;
import org.cloudfoundry.credhub.util.DatabaseProfileResolver;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import java.security.Key;
import java.util.HashMap;
import java.util.UUID;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@RunWith(JUnit4.class)
@SpringBootTest
@ActiveProfiles(value = "unit-test", resolver = DatabaseProfileResolver.class)
public class KeyUsageControllerTest {

  private MockMvc mockMvc;
  CredentialVersionDataService credentialVersionDataService;
  EncryptionKeyCanaryDataService encryptionKeyCanaryDataService;
  EncryptionKeySet keySet;

  @Before
  public void beforeEach() {
    credentialVersionDataService = mock(CredentialVersionDataService.class);
    keySet = new EncryptionKeySet();
    encryptionKeyCanaryDataService = mock(EncryptionKeyCanaryDataService.class);
    final KeyUsageController keyUsageController = new KeyUsageController(credentialVersionDataService,
        keySet);

    mockMvc = MockMvcBuilders
        .standaloneSetup(keyUsageController)
        .alwaysDo(print())
        .build();
  }

  @Test
  public void getKeyUsages_getsKeyDistributionAcrossActiveInactiveAndUnknownEncryptionKeys()
      throws Exception {
    final UUID activeKey = UUID.randomUUID();
    final UUID knownKey = UUID.randomUUID();
    final UUID unknownKey = UUID.randomUUID();

    HashMap<UUID, Long> countByEncryptionKey = new HashMap<>();
    countByEncryptionKey.put(activeKey, 200L);
    countByEncryptionKey.put(knownKey, 10L);
    countByEncryptionKey.put(unknownKey, 5L);

    keySet.add(new EncryptionKey(mock(EncryptionService.class), activeKey, mock(Key.class)));
    keySet.add(new EncryptionKey(mock(EncryptionService.class), knownKey, mock(Key.class)));
    keySet.setActive(activeKey);
    when(credentialVersionDataService.countByEncryptionKey()).thenReturn(countByEncryptionKey);

    mockMvc.perform(get("/api/v1/key-usage"))
        .andExpect(status().isOk())
        .andExpect(content().contentTypeCompatibleWith(MediaType.APPLICATION_JSON))
        .andExpect(jsonPath("$.active_key").value(200))
        .andExpect(jsonPath("$.inactive_keys").value(10))
        .andExpect(jsonPath("$.unknown_keys").value(5));
  }

}

File: backends/credhub/src/test/kotlin/org/cloudfoundry/credhub/controllers/v1/KeyUsageControllerTest.kt
package org.cloudfoundry.credhub.controllers.v1

import org.cloudfoundry.credhub.keyusage.KeyUsageController
import org.cloudfoundry.credhub.services.CredentialVersionDataService
import org.cloudfoundry.credhub.services.InternalEncryptionService
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.springframework.http.MediaType
import org.springframework.test.context.junit4.SpringRunner
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.result.MockMvcResultHandlers.print
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.content
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import org.springframework.test.web.servlet.setup.StandaloneMockMvcBuilder
import java.security.Key
import java.util.*

@RunWith(SpringRunner::class)
class KeyUsageControllerTest {

    private lateinit var credentialVersionDataService: CredentialVersionDataService
    private lateinit var keySet: org.cloudfoundry.credhub.services.EncryptionKeySet
    private lateinit var mockMvc: MockMvc

    @Before
    fun beforeEach() {
        credentialVersionDataService = mock(CredentialVersionDataService::class.java)
        keySet = org.cloudfoundry.credhub.services.EncryptionKeySet()

        val keyUsageController = KeyUsageController(credentialVersionDataService,
            keySet)

        mockMvc = MockMvcBuilders
            .standaloneSetup(keyUsageController)
            .alwaysDo<StandaloneMockMvcBuilder>(print())
            .build()
    }

    @Test
    fun `GET key usages gets key distribution across active, inactive, and unknown encryption keys`() {
        val activeKey = UUID.randomUUID()
        val knownKey = UUID.randomUUID()
        val unknownKey = UUID.randomUUID()

        val countByEncryptionKey = HashMap<UUID, Long>()
        countByEncryptionKey[activeKey] = 200L
        countByEncryptionKey[knownKey] = 10L
        countByEncryptionKey[unknownKey] = 5L

        keySet.add(org.cloudfoundry.credhub.services.EncryptionKey(mock(InternalEncryptionService::class.java), activeKey, mock(Key::class.java), "key-name"))
        keySet.add(org.cloudfoundry.credhub.services.EncryptionKey(mock(InternalEncryptionService::class.java), knownKey, mock(Key::class.java), "key-name"))
        keySet.setActive(activeKey)
        `when`(credentialVersionDataService.countByEncryptionKey()).thenReturn(countByEncryptionKey)

        mockMvc.perform(get(KeyUsageController.ENDPOINT))
            .andExpect(status().isOk)
            .andExpect(content().contentTypeCompatibleWith(MediaType.APPLICATION_JSON))
            .andExpect(jsonPath("$.active_key").value(200))
            .andExpect(jsonPath("$.inactive_keys").value(10))
            .andExpect(jsonPath("$.unknown_keys").value(5))
    }
}
